distributed/worker_memory.py

```diff
@@ -0,0 +1,384 @@
+"""Encapsulated manager for in-memory tasks on a worker.
+
+This module covers:
+- spill/unspill data depending on the 'distributed.worker.memory.target' threshold
+- spill/unspill data depending on the 'distributed.worker.memory.spill' threshold
+- pause/unpause the worker depending on the 'distributed.worker.memory.pause' threshold
+- kill the worker depending on the 'distributed.worker.memory.terminate' threshold
+
+This module does *not* cover:
+- Changes in behaviour in Worker, Scheduler, task stealing, Active Memory Manager, etc.
+  caused by the Worker being in paused status
+- Worker restart after it's been killed
+- Scheduler-side heuristics regarding memory usage, e.g. the Active Memory Manager
+
+See also:
+- :mod:`distributed.spill`, which implements the spill-to-disk mechanism and is wrapped
+  around by this module. Unlike this module, :mod:`distributed.spill` is agnostic to the
+  Worker.
+- :mod:`distributed.active_memory_manager`, which runs on the scheduler side
+"""
+from __future__ import annotations
+
+import asyncio
+import logging
+import os
+import sys
+import warnings
+from collections.abc import Callable, MutableMapping
+from contextlib import suppress
+from functools import partial
+from typing import TYPE_CHECKING, Any, Container, Literal
+
+import psutil
+from tornado.ioloop import PeriodicCallback
+
+import dask.config
+from dask.system import CPU_COUNT
+from dask.utils import format_bytes, parse_bytes, parse_timedelta
+
+from . import system
+from .core import Status
+from .spill import SpillBuffer
+from .utils_perf import ThrottledGC
+
+if TYPE_CHECKING:
+    # Circular imports
+    from .nanny import Nanny
+    from .worker import Worker
+
+logger = logging.getLogger(__name__)
+
+
+class WorkerMemoryManager:
+    data: MutableMapping[str, Any]  # {task key: task payload}
+    memory_limit: int | None
+    memory_target_fraction: float | Literal[False]
+    memory_spill_fraction: float | Literal[False]
+    memory_pause_fraction: float | Literal[False]
+    max_spill: int | Literal[False]
+    memory_monitor_interval: float
+    _memory_monitoring: bool
+    _throttled_gc: ThrottledGC
+
+    def __init__(
+        self,
+        worker: Worker,
+        *,
+        memory_limit: str | float = "auto",
+        # This should be None most of the times, short of a power user replacing the
+        # SpillBuffer with their own custom dict-like
+        data: (
+            MutableMapping[str, Any]  # pre-initialised
+            | Callable[[], MutableMapping[str, Any]]  # constructor
+            | tuple[
+                Callable[..., MutableMapping[str, Any]], dict[str, Any]
+            ]  # (constructor, kwargs to constructor)
+            | None  # create internally
+        ) = None,
```

Comment:
This seems to me like a feature for power user that is as powerful as obscure, and I would *love* to hear from somebody who actually uses it to understand their use case!

Reply:
This is used by Dask-CUDA. See https://github.com/dask/distributed/issues/5909

Reply:
I wonder if `ManualEvictProto` could be used in this typed declaration, but I have no strong opinion, as I'm not sure whether it would even be possible.

Reply:
No, because the type declaration says what the parameter must be; ManualEvictProto is an additional, optional interface to the MutableMapping which unlocks extra features.